### PR TITLE
KMeansClustering Implementation

### PR DESCRIPTION
**Pull request by menuka94**
Example Request:
```json
{
   "type":"K_MEANS_CLUSTERING",
   "kMeansClusteringRequest":{
      "clusterCount":10,
      "maxIterations":100,
      "resolution":"County",
      "features":[
         "total_population",
         "median_household_income"
      ]
   }
}
```

Example (streaming) response:
```json
{
   "kMeansClusteringResponse":{
      "gisJoin":"G3701330",
      "prediction":9
   }
}
```

### CODE REVIEW

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java [hunk 1]
```diff
@@ -1,23 +1,169 @@
...
+        KMeansClusteringRequest kMeansClusteringRequest = request.getKMeansClusteringRequest();
+        int k = kMeansClusteringRequest.getClusterCount();
+        int maxIterations = kMeansClusteringRequest.getMaxIterations();
+        ArrayList<String> features = new ArrayList<>(kMeansClusteringRequest.getFeaturesList());
+        log.info("\tk: " + k);
+        log.info("\tmaxIterations: " + maxIterations);
+        log.info("\tfeatures:");
+        for (String feature : features) {
+            log.info("\t\tfeature: " + feature);
+        }
```
Comment (hamersaw) [inline src/main/java/org/sustain/modeling/ClusteringQueryHandler.java:145]: try to keep log request to a single line. it ensures when parsing logs that messages from other requests don't inject themselves into the sequence.

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java [hunk 2]
```diff
@@ -1,23 +1,169 @@
...
+        int maxIterations = request.getKMeansClusteringRequest().getMaxIterations();
+        Seq<String> features = convertListToSeq(featuresList);
+
+        Dataset<Row> selectedFeatures = collection.select(Constants.GIS_JOIN, features);
+
+        // KMeans Clustering
+        VectorAssembler assembler =
+                new VectorAssembler().setInputCols(featuresList.toArray(new String[0])).setOutputCol("features");
+        Dataset<Row> featureDF = assembler.transform(selectedFeatures);
+        KMeans kmeans = new KMeans().setK(k).setSeed(1L);
```
Comment (hamersaw) [inline src/main/java/org/sustain/modeling/ClusteringQueryHandler.java:79]: should the seed value be a parameter on KMeansClusteringRequest?

**Review by hamersaw — APPROVED** [review]

**Comment by hamersaw** [comment]
We're going to need to standardize protobuf definitions for collection features and resolutions. Maybe something to push off until next week?

**Comment by drg101** [comment]
Shouldn't the request define which collections, and which features within those collections to perform the model on? That's how the model catalogue sample had been set up.

**Comment by hamersaw** [comment]
We had toyed with ideas about how to define which features are available through modelling. The difficulty is that different models allow for different features. Additionally, there are different types of features available for each models (e.x. training and target features for regression models). I like the idea of keeping the model catalogue restricted to model definitions and parameters and having some other mechanism for defining model features. As for grpc communication this protobuf and the linear regression definition have differing constructs for feature definition. They should certainly be defined at the ModelRequest level, because all features will have models of some type. I like  something like

message ModelRequest {
  repeated Feature features = x;
}

message Feature {
  string collection = 1;
  string feature = 2;
  string role = 3;
}

and then we have some parsing to do on the backend to validate the set of features for the model definition. Like I said, I think this warrants a longer discussion. Perhaps tackling it after the PI meeting would be easier?

**Comment by drg101** [comment]
I like that setup, I think the ability to perform these modeling operations on specific parts of collections is very important, as many of these collections contain lots of junk that will not be useful for any sort of modeling.

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java (hunk 2 above)
Comment (menuka94) [reply src/main/java/org/sustain/modeling/ClusteringQueryHandler.java:79]: Technically, yes. But I couldn't find the correct range for the seed in the documentation. So I decided to keep it hard-coded for now.

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java (hunk 1 above)
Comment (menuka94) [reply src/main/java/org/sustain/modeling/ClusteringQueryHandler.java]: Done

**Review by inf0rmatiker — APPROVED** [review]

**Pull request merged**